src/shogun/modelselection/parameter_tree.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_PARAMETER_TREE_H
+#define SHOGUN_PARAMETER_TREE_H
+
+#include <shogun/base/AnyParameter.h>
+#include <shogun/base/SGObject.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/type_case.h>
+#include <shogun/machine/Machine.h>
+#include <unordered_set>
+
+namespace shogun
+{
+
+	class ParameterNode : public CSGObject
```

Comment:
It would be good to have some kind of high level explain of the parts involved in here and how they play together, in some doxygen comment. For future development of the stuff. Not now, but definitely later